library/ceph_volume.py

```diff
@@ -250,6 +235,114 @@ def run_module():
     module.exit_json(**result)
 
 
+def zap_devices(module):
+    """
+    Will run 'ceph-volume lvm zap' on all devices, lvs and partitions
+    used to create the OSD. The --destroy flag is always passed so that
+    if an OSD was originally created with a raw device or partition for
+    'data' then any lvs that were created by ceph-volume are removed.
+    """
+    data = module.params['data']
+    data_vg = module.params.get('data_vg', None)
+    journal = module.params.get('journal', None)
+    journal_vg = module.params.get('journal_vg', None)
+    db = module.params.get('db', None)
+    db_vg = module.params.get('db_vg', None)
+    wal = module.params.get('wal', None)
+    wal_vg = module.params.get('wal_vg', None)
+
+    base_zap_cmd = [
+        'ceph-volume',
+        'lvm',
+        'zap',
+        # for simplicity always --destroy. It will be needed
+        # for raw devices and will noop for lvs.
+        '--destroy',
+    ]
+
+    commands = []
+
+    data = get_data(data, data_vg)
+
+    commands.append(base_zap_cmd + [data])
+
+    if journal:
+        journal = get_journal(journal, journal_vg)
+        commands.append(base_zap_cmd + [journal])
+
+    if db:
+        db = get_db(db, db_vg)
+        commands.append(base_zap_cmd + [db])
+
+    if wal:
+        wal = get_wal(wal, wal_vg)
+        commands.append(base_zap_cmd + [wal])
+
+    result = dict(
+        changed=True,
+        rc=0,
+    )
+    command_results = []
+    for cmd in commands:
+        startd = datetime.datetime.now()
+
+        rc, out, err = module.run_command(cmd, encoding=None)
+
+        endd = datetime.datetime.now()
+        delta = endd - startd
+
+        cmd_result = dict(
+            cmd=cmd,
+            stdout_lines=out.split("\n"),
+            stderr_lines=err.split("\n"),
+            rc=rc,
+            start=str(startd),
+            end=str(endd),
+            delta=str(delta),
+        )
+
+        if rc != 0:
+            module.fail_json(msg='non-zero return code', **cmd_result)
+
+        command_results.append(cmd_result)
+
+    result["commands"] = command_results
+
+    module.exit_json(**result)
+
+
+def run_module():
+    module_args = dict(
+        cluster=dict(type='str', required=False, default='ceph'),
+        objectstore=dict(type='str', required=False, choices=['bluestore', 'filestore']),
+        action=dict(type='str', required=False, choices=['create', 'zap'], default='create'),
+        data=dict(type='str', required=True),
```

Comment:
You need to add `default=` so you don't need to run  `module.params.get('data_vg', None)` and apply a default when trying the get the variable.

Reply:
There would be no way to set a default for ``data_vg`` that made any sense. I'm unsure if Ansible will accept a default of ``None`` for a ``type='str'``, but I guess I could try.

Something like ``module.params.get('data_vg', None)`` is a very common pattern in python, are you not liking the extra bit of verbosity there?

Reply:
Yes I know it's common, I have the impression that `module.params.get('data_vg', None)` achieves the same. That's what I observed but I might be wrong.

If that happens not being the case, then I'm okay with your code.

Reply:
Yeah, we don't gain anything by making that change. The ``, None)`` bit there is actually not needed but I like to be verbose. A ``.get()`` will return ``None`` if it doesn't find the key you're looking for.